libfreerdp/core/gateway/websocket.c

```diff
@@ -0,0 +1,497 @@
+/**
+ * FreeRDP: A Remote Desktop Protocol Implementation
+ * Websocket Framing
+ *
+ * Copyright 2023 Michael Saxl <mike@mwsys.mine.bz>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "websocket.h"
+
+#define TAG FREERDP_TAG("core.gateway.websocket")
+
+BOOL websocket_write_wstream(BIO* bio, wStream* sPacket, WEBSOCKET_OPCODE opcode)
+{
+	size_t len;
+	size_t fullLen;
+	int status;
+	wStream* sWS;
+
+	uint32_t maskingKey;
+
+	size_t streamPos;
+
+	len = Stream_Length(sPacket);
+	Stream_SetPosition(sPacket, 0);
+
+	if (len > INT_MAX)
+		return FALSE;
+
+	if (len < 126)
+		fullLen = len + 6; /* 2 byte "mini header" + 4 byte masking key */
+	else if (len < 0x10000)
+		fullLen = len + 8; /* 2 byte "mini header" + 2 byte length + 4 byte masking key */
+	else
+		fullLen = len + 14; /* 2 byte "mini header" + 8 byte length + 4 byte masking key */
+
+	sWS = Stream_New(NULL, fullLen);
+	if (!sWS)
+		return FALSE;
+
+	winpr_RAND((BYTE*)&maskingKey, 4);
+
+	Stream_Write_UINT8(sWS, WEBSOCKET_FIN_BIT | opcode);
+	if (len < 126)
+		Stream_Write_UINT8(sWS, len | WEBSOCKET_MASK_BIT);
+	else if (len < 0x10000)
+	{
+		Stream_Write_UINT8(sWS, 126 | WEBSOCKET_MASK_BIT);
+		Stream_Write_UINT16_BE(sWS, len);
+	}
+	else
+	{
+		Stream_Write_UINT8(sWS, 127 | WEBSOCKET_MASK_BIT);
+		Stream_Write_UINT32_BE(sWS, 0); /* payload is limited to INT_MAX */
+		Stream_Write_UINT32_BE(sWS, len);
+	}
+	Stream_Write_UINT32(sWS, maskingKey);
+
+	/* mask as much as possible with 32bit access */
+	for (streamPos = 0; streamPos + 4 <= len; streamPos += 4)
+	{
+		uint32_t data;
+		Stream_Read_UINT32(sPacket, data);
+		Stream_Write_UINT32(sWS, data ^ maskingKey);
+	}
+
+	/* mask the rest byte by byte */
+	for (; streamPos < len; streamPos++)
+	{
+		BYTE data;
+		BYTE* partialMask = ((BYTE*)&maskingKey) + (streamPos % 4);
+		Stream_Read_UINT8(sPacket, data);
+		Stream_Write_UINT8(sWS, data ^ *partialMask);
+	}
+
+	Stream_SealLength(sWS);
+
+	ERR_clear_error();
+	status = BIO_write(bio, Stream_Buffer(sWS), Stream_Length(sWS));
```

Comment:
best `WINPR_ASSERT(bio);` here

Reply:
I've put the asserts on the beginning of each function. There is no reason to try doing things if bio isn't there, so fail early